synapse/storage/prepare_database.py

```diff
@@ -59,6 +58,28 @@ class UpgradeDatabaseException(PrepareDatabaseException):
 )
 
 
+@attr.s
+class _SchemaState:
+    current_version: int = attr.ib()
+    """The current schema version of the database"""
+
+    compat_version: Optional[int] = attr.ib()
+    """The SCHEMA_VERSION of the oldest version of Synapse for this database
+
+    If this is None, we have an old version of the database without the necessary
+    table.
+    """
+
+    applied_deltas: Collection[str] = attr.ib(factory=tuple)
+    """Any delta files for `current_version` which have already been applied"""
+
+    upgraded: bool = attr.ib(default=False)
```

Comment:
(Fwiw you could remove all the `attr.ib()` here and use `auto_attribs=True`)

Reply:
yeah, I think I didn't know about `auto_attribs` when I first wrote this code...